syft/pkg/cataloger/golang/parse_go_bin.go

```diff
@@ -39,57 +67,121 @@ func newGoBinaryPackage(name, version, h1Digest, goVersion, architecture string,
 	return p
 }
 
-func parseGoBin(location source.Location, reader io.ReadCloser, opener exeOpener) (pkgs []pkg.Package, err error) {
-	var exes []exe
-	// it has been found that there are stdlib paths within openExe that can panic. We want to prevent this behavior
-	// bubbling up and halting execution. For this reason we try to recover from any panic and return an error.
-	defer func() {
-		if r := recover(); r != nil {
-			err = fmt.Errorf("recovered from panic while parse go binary at %q: %+v", location.RealPath, r)
+// getArchs finds a binary architecture by two ways:
+// 1) reading build info from binaries compiled by go1.18+
+// 2) reading file headers from binaries compiled by < go1.18
+func getArchs(readers []io.ReaderAt, builds []*debug.BuildInfo) []string {
+	if len(readers) != len(builds) {
+		log.Errorf("golang cataloger: bin parsing: number of builds and readers doesn't match")
+		return nil
+	}
+
+	archs := make([]string, len(builds))
+	for i, build := range builds {
+		archs[i] = getGOARCH(build.Settings)
+	}
+
+	// if architecture was found via build settings return
+	if archs[0] != "" {
+		return archs
+	}
+
+	for i, r := range readers {
+		a, err := getGOARCHFromBin(r)
+		if err != nil || a == "" {
+			log.Warnf("golang cataloger: bin parsing: getting arch from binary: %v", err)
```

Comment:
we're logging the error even though the `if` case may show that as `nil`. We should have two different log cases for this instead.

Reply:
fixed